File: app/src/main/java/com/kunzisoft/keepass/adapters/NodeAdapter.java
/*
 * Copyright 2018 Jeremy Jamet / Kunzisoft.
 *
 * This file is part of KeePass DX.
 *
 *  KeePass DX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePass DX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePass DX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.adapters;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.v7.util.SortedList;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.util.SortedListAdapterCallback;
import android.util.Log;
import android.view.*;
import android.widget.Toast;
import com.kunzisoft.keepass.R;
import com.kunzisoft.keepass.app.App;
import com.kunzisoft.keepass.database.SortNodeEnum;
import com.kunzisoft.keepass.database.element.*;
import com.kunzisoft.keepass.settings.PreferencesUtil;
import com.kunzisoft.keepass.utils.Util;

public class NodeAdapter extends RecyclerView.Adapter<BasicViewHolder> {
    private static final String TAG = NodeAdapter.class.getName();

    private SortedList<NodeVersioned> nodeSortedList;

    private Context context;
    private LayoutInflater inflater;
    private MenuInflater menuInflater;
    private float textSize;
    private float subtextSize;
    private float iconSize;
    private SortNodeEnum listSort;
    private boolean groupsBeforeSort;
    private boolean ascendingSort;
    private boolean showUsernames;

    private NodeClickCallback nodeClickCallback;
    private NodeMenuListener nodeMenuListener;
    private boolean activateContextMenu;
    private boolean readOnly;
    private boolean isASearchResult;

    private Database database;

    private int iconGroupColor;
    private int iconEntryColor;

    /**
     * Create node list adapter with contextMenu or not
     * @param context Context to use
     */
    public NodeAdapter(final Context context, MenuInflater menuInflater) {
        this.inflater = LayoutInflater.from(context);
        this.menuInflater = menuInflater;
        this.context = context;
        assignPreferences();
        this.activateContextMenu = false;
        this.readOnly = false;
        this.isASearchResult = false;

        this.nodeSortedList = new SortedList<>(NodeVersioned.class, new SortedListAdapterCallback<NodeVersioned>(this) {
            @Override public int compare(NodeVersioned item1, NodeVersioned item2) {
                return listSort.getNodeComparator(ascendingSort, groupsBeforeSort).compare(item1, item2);
            }

            @Override public boolean areContentsTheSame(NodeVersioned oldItem, NodeVersioned newItem) {
				return oldItem.getTitle().equals(newItem.getTitle())
						&& oldItem.getIcon().equals(newItem.getIcon());
            }

            @Override public boolean areItemsTheSame(NodeVersioned item1, NodeVersioned item2) {
                return item1.equals(item2);
            }
        });

        // Database
        this.database = App.getDB();

        // Retrieve the color to tint the icon
        int[] attrTextColorPrimary = {android.R.attr.textColorPrimary};
        TypedArray taTextColorPrimary = context.getTheme().obtainStyledAttributes(attrTextColorPrimary);
        this.iconGroupColor = taTextColorPrimary.getColor(0, Color.BLACK);
        taTextColorPrimary.recycle();
        int[] attrTextColor = {android.R.attr.textColor}; // In two times to fix bug compilation
        TypedArray taTextColor = context.getTheme().obtainStyledAttributes(attrTextColor);
        this.iconEntryColor = taTextColor.getColor(0, Color.BLACK);
        taTextColor.recycle();
    }

    public void setReadOnly(boolean readOnly) {
        this.readOnly = readOnly;
    }

    public void setIsASearchResult(boolean isASearchResult) {
        this.isASearchResult = isASearchResult;
    }

    public void setActivateContextMenu(boolean activate) {
        this.activateContextMenu = activate;
    }

    private void assignPreferences() {
        float textSizeDefault = Util.getListTextDefaultSize(context);
        this.textSize = PreferencesUtil.getListTextSize(context);
        this.subtextSize = context.getResources().getInteger(R.integer.list_small_size_default)
                * textSize / textSizeDefault;
        // Retrieve the icon size
        float iconDefaultSize = context.getResources().getDimension(R.dimen.list_icon_size_default);
        this.iconSize = iconDefaultSize * textSize / textSizeDefault;
        this.listSort = PreferencesUtil.getListSort(context);
        this.groupsBeforeSort = PreferencesUtil.getGroupsBeforeSort(context);
        this.ascendingSort = PreferencesUtil.getAscendingSort(context);
        this.showUsernames = PreferencesUtil.showUsernamesListEntries(context);
    }

    /**
     * Rebuild the list by clear and build children from the group
     */
    public void rebuildList(GroupVersioned group) {
        this.nodeSortedList.clear();
        assignPreferences();
        // TODO verify sort
        try {
            this.nodeSortedList.addAll(group.getChildEntriesWithoutMetaStream());
        } catch (Exception e) {
            Log.e(TAG, "Can't add node elements to the list", e);
            Toast.makeText(context, "Can't add node elements to the list : " + e.getMessage(), Toast.LENGTH_LONG).show();
        }
    }

    /**
     * Determine if the adapter contains or not any element
     * @return true if the list is empty
     */
    public boolean isEmpty() {
        return nodeSortedList.size() <= 0;
    }

    /**
     * Add a node to the list
     * @param node Node to add
     */
    public void addNode(NodeVersioned node) {
        nodeSortedList.add(node);
    }

    /**
     * Remove a node in the list
     * @param node Node to delete
     */
    public void removeNode(NodeVersioned node) {
        nodeSortedList.remove(node);
    }

    /**
     * Update a node in the list
     * @param oldNode Node before the update
     * @param newNode Node after the update
     */
    public void updateNode(NodeVersioned oldNode, NodeVersioned newNode) {
        nodeSortedList.beginBatchedUpdates();
        nodeSortedList.remove(oldNode);
        nodeSortedList.add(newNode);
        nodeSortedList.endBatchedUpdates();
    }

    /**
     * Notify a change sort of the list
     */
    public void notifyChangeSort(SortNodeEnum sortNodeEnum, boolean ascending, boolean groupsBefore) {
        this.listSort = sortNodeEnum;
        this.ascendingSort = ascending;
        this.groupsBeforeSort = groupsBefore;
    }

    @Override
    public int getItemViewType(int position) {
        return nodeSortedList.get(position).getType().ordinal();
    }

    @NonNull
    @Override
    public BasicViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        BasicViewHolder basicViewHolder;
        View view;
        if (viewType == Type.GROUP.ordinal()) {
            view = inflater.inflate(R.layout.list_nodes_group, parent, false);
            basicViewHolder = new GroupViewHolder(view);
        } else {
            view = inflater.inflate(R.layout.list_nodes_entry, parent, false);
            basicViewHolder = new EntryViewHolder(view);
        }
        return basicViewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull BasicViewHolder holder, int position) {
        NodeVersioned subNode = nodeSortedList.get(position);
        // Assign image
        int iconColor = Color.BLACK;
        switch (subNode.getType()) {
            case GROUP:
                iconColor = iconGroupColor;
                break;
            case ENTRY:
                iconColor = iconEntryColor;
                break;
        }
        database.getDrawFactory().assignDatabaseIconTo(context, holder.icon, subNode.getIcon(), iconColor);
        // Assign text
        holder.text.setText(subNode.getTitle());
        // Assign click
        holder.container.setOnClickListener(
                new OnNodeClickListener(subNode));
        // Context menu
        if (activateContextMenu) {
            holder.container.setOnCreateContextMenuListener(
                    new ContextMenuBuilder(subNode, nodeMenuListener, readOnly));
        }

        // Add username
        holder.subText.setText("");
        holder.subText.setVisibility(View.GONE);
        if (subNode.getType().equals(Type.ENTRY)) {
            EntryVersioned entry = (EntryVersioned) subNode;

            database.startManageEntry(entry);

            holder.text.setText(entry.getVisualTitle());

            String username = entry.getUsername();
            if (showUsernames && !username.isEmpty()) {
                holder.subText.setVisibility(View.VISIBLE);
                holder.subText.setText(username);
            }

            database.stopManageEntry(entry);
        }

        // Assign image and text size
        // Relative size of the icon
        holder.icon.getLayoutParams().height = ((int) iconSize);
        holder.icon.getLayoutParams().width = ((int) iconSize);
        holder.text.setTextSize(textSize);
        holder.subText.setTextSize(subtextSize);
    }

    @Override
    public int getItemCount() {
        return nodeSortedList.size();
    }

    /**
     * Assign a listener when a node is clicked
     */
    public void setOnNodeClickListener(NodeClickCallback nodeClickCallback) {
        this.nodeClickCallback = nodeClickCallback;
    }

    /**
     * Assign a listener when an element of menu is clicked
     */
    public void setNodeMenuListener(NodeMenuListener nodeMenuListener) {
        this.nodeMenuListener = nodeMenuListener;
    }

    /**
     * Callback listener to redefine to do an action when a node is click
     */
    public interface NodeClickCallback {
        void onNodeClick(NodeVersioned node);
    }

    /**
     * Menu listener to redefine to do an action in menu
     */
    public interface NodeMenuListener {
        boolean onOpenMenuClick(NodeVersioned node);
        boolean onEditMenuClick(NodeVersioned node);
        boolean onCopyMenuClick(NodeVersioned node);
        boolean onMoveMenuClick(NodeVersioned node);
        boolean onDeleteMenuClick(NodeVersioned node);
    }

    /**
     * Utility class for node listener
     */
    private class OnNodeClickListener implements View.OnClickListener {
        private NodeVersioned node;

        OnNodeClickListener(NodeVersioned node) {
            this.node = node;
        }

        @Override
        public void onClick(View v) {
            if (nodeClickCallback != null)
                nodeClickCallback.onNodeClick(node);
        }
    }

    /**
     * Utility class for menu listener
     */
    private class ContextMenuBuilder implements View.OnCreateContextMenuListener {

        private NodeVersioned node;
        private NodeMenuListener menuListener;
        private boolean readOnly;

        ContextMenuBuilder(NodeVersioned node, NodeMenuListener menuListener, boolean readOnly) {
            this.menuListener = menuListener;
            this.node = node;
            this.readOnly = readOnly;
        }

        @Override
        public void onCreateContextMenu(ContextMenu contextMenu, View view, ContextMenu.ContextMenuInfo contextMenuInfo) {
            menuInflater.inflate(R.menu.node_menu, contextMenu);

            // Opening
            MenuItem menuItem = contextMenu.findItem(R.id.menu_open);
            menuItem.setOnMenuItemClickListener(mOnMyActionClickListener);

			Database database = App.getDB();

            // Edition
            if (readOnly || node.equals(database.getRecycleBin())) {
                contextMenu.removeItem(R.id.menu_edit);
            } else {
                menuItem = contextMenu.findItem(R.id.menu_edit);
                menuItem.setOnMenuItemClickListener(mOnMyActionClickListener);
            }

            // Copy (not for group)
            if (readOnly
                    || isASearchResult
                    || node.equals(database.getRecycleBin())
                    || node.getType().equals(Type.GROUP)) {
                // TODO COPY For Group
                contextMenu.removeItem(R.id.menu_copy);
            } else {
                menuItem = contextMenu.findItem(R.id.menu_copy);
                menuItem.setOnMenuItemClickListener(mOnMyActionClickListener);
            }

            // Move
            if (readOnly
                    || isASearchResult
                    || node.equals(database.getRecycleBin())) {
                contextMenu.removeItem(R.id.menu_move);
            } else {
                menuItem = contextMenu.findItem(R.id.menu_move);
                menuItem.setOnMenuItemClickListener(mOnMyActionClickListener);
            }

            // Deletion
            if (readOnly || node.equals(database.getRecycleBin())) {
                contextMenu.removeItem(R.id.menu_delete);
            } else {
                menuItem = contextMenu.findItem(R.id.menu_delete);
                menuItem.setOnMenuItemClickListener(mOnMyActionClickListener);
            }
        }

        private MenuItem.OnMenuItemClickListener mOnMyActionClickListener = new MenuItem.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                if (menuListener == null)
                    return false;
                switch ( item.getItemId() ) {
                    case R.id.menu_open:
                        return menuListener.onOpenMenuClick(node);
                    case R.id.menu_edit:
                        return menuListener.onEditMenuClick(node);
                    case R.id.menu_copy:
                        return menuListener.onCopyMenuClick(node);
                    case R.id.menu_move:
                        return menuListener.onMoveMenuClick(node);
                    case R.id.menu_delete:
                        return menuListener.onDeleteMenuClick(node);
                    default:
                        return false;
                }
            }
        };
    }
}

File: app/src/main/java/com/kunzisoft/keepass/adapters/NodeAdapter.kt
/*
 * Copyright 2019 Jeremy Jamet / Kunzisoft.
 *
 * This file is part of KeePassDX.
 *
 *  KeePassDX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePassDX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePassDX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.adapters

import android.content.Context
import android.graphics.Color
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SortedList
import androidx.recyclerview.widget.SortedListAdapterCallback
import com.kunzisoft.keepass.R
import com.kunzisoft.keepass.database.element.Database
import com.kunzisoft.keepass.database.element.Entry
import com.kunzisoft.keepass.database.element.Group
import com.kunzisoft.keepass.database.element.SortNodeEnum
import com.kunzisoft.keepass.database.element.node.Node
import com.kunzisoft.keepass.database.element.node.NodeVersionedInterface
import com.kunzisoft.keepass.database.element.node.Type
import com.kunzisoft.keepass.database.element.template.TemplateField
import com.kunzisoft.keepass.otp.OtpElement
import com.kunzisoft.keepass.otp.OtpType
import com.kunzisoft.keepass.settings.PreferencesUtil
import com.kunzisoft.keepass.timeout.ClipboardHelper
import com.kunzisoft.keepass.view.setTextSize
import com.kunzisoft.keepass.view.strikeOut
import java.util.*

/**
 * Create node list adapter with contextMenu or not
 * @param context Context to use
 */
class NodeAdapter (private val context: Context,
                   private val database: Database)
    : RecyclerView.Adapter<NodeAdapter.NodeViewHolder>() {

    private var mNodeComparator: Comparator<NodeVersionedInterface<Group>>? = null
    private val mNodeSortedListCallback: NodeSortedListCallback
    private val mNodeSortedList: SortedList<Node>
    private val mInflater: LayoutInflater = LayoutInflater.from(context)

    private var mCalculateViewTypeTextSize = Array(2) { true } // number of view type
    private var mTextSizeUnit: Int = TypedValue.COMPLEX_UNIT_PX
    private var mPrefSizeMultiplier: Float = 0F
    private var mSubtextDefaultDimension: Float = 0F
    private var mInfoTextDefaultDimension: Float = 0F
    private var mNumberChildrenTextDefaultDimension: Float = 0F
    private var mIconDefaultDimension: Float = 0F

    private var mShowUserNames: Boolean = true
    private var mShowNumberEntries: Boolean = true
    private var mShowOTP: Boolean = false
    private var mShowUUID: Boolean = false
    private var mEntryFilters = arrayOf<Group.ChildFilter>()

    private var mActionNodesList = LinkedList<Node>()
    private var mNodeClickCallback: NodeClickCallback? = null
    private var mClipboardHelper = ClipboardHelper(context)

    @ColorInt
    private val mContentSelectionColor: Int
    @ColorInt
    private val mIconGroupColor: Int
    @ColorInt
    private val mIconEntryColor: Int

    /**
     * Determine if the adapter contains or not any element
     * @return true if the list is empty
     */
    val isEmpty: Boolean
        get() = mNodeSortedList.size() <= 0

    init {
        this.mIconDefaultDimension = context.resources.getDimension(R.dimen.list_icon_size_default)

        assignPreferences()

        this.mNodeSortedListCallback = NodeSortedListCallback()
        this.mNodeSortedList = SortedList(Node::class.java, mNodeSortedListCallback)

        // Color of content selection
        this.mContentSelectionColor = ContextCompat.getColor(context, R.color.white)
        // Retrieve the color to tint the icon
        val taTextColorPrimary = context.theme.obtainStyledAttributes(intArrayOf(android.R.attr.textColorPrimary))
        this.mIconGroupColor = taTextColorPrimary.getColor(0, Color.BLACK)
        taTextColorPrimary.recycle()
        // In two times to fix bug compilation
        val taTextColor = context.theme.obtainStyledAttributes(intArrayOf(android.R.attr.textColor))
        this.mIconEntryColor = taTextColor.getColor(0, Color.BLACK)
        taTextColor.recycle()
    }

    private fun assignPreferences() {
        this.mPrefSizeMultiplier = PreferencesUtil.getListTextSize(context)

        notifyChangeSort(
                PreferencesUtil.getListSort(context),
                        SortNodeEnum.SortNodeParameters(
                            PreferencesUtil.getAscendingSort(context),
                            PreferencesUtil.getGroupsBeforeSort(context),
                            PreferencesUtil.getRecycleBinBottomSort(context)
                        )
                )

        this.mShowUserNames = PreferencesUtil.showUsernamesListEntries(context)
        this.mShowNumberEntries = PreferencesUtil.showNumberEntries(context)
        this.mShowOTP = PreferencesUtil.showOTPToken(context)
        this.mShowUUID = PreferencesUtil.showUUID(context)

        this.mEntryFilters = Group.ChildFilter.getDefaults(context)

        // Reinit textSize for all view type
        mCalculateViewTypeTextSize.forEachIndexed { index, _ -> mCalculateViewTypeTextSize[index] = true }
    }

    /**
     * Rebuild the list by clear and build children from the group
     */
    fun rebuildList(group: Group) {
        assignPreferences()
        mNodeSortedList.replaceAll(group.getFilteredChildren(mEntryFilters))
    }

    private inner class NodeSortedListCallback: SortedListAdapterCallback<Node>(this) {
        override fun compare(item1: Node, item2: Node): Int {
            return mNodeComparator!!.compare(item1, item2)
        }

        override fun areContentsTheSame(oldItem: Node, newItem: Node): Boolean {
            var typeContentTheSame = true
            if (oldItem is Entry && newItem is Entry) {
                typeContentTheSame = oldItem.getVisualTitle() == newItem.getVisualTitle()
                        && oldItem.username == newItem.username
                        && oldItem.getOtpElement() == newItem.getOtpElement()
                        && oldItem.containsAttachment() == newItem.containsAttachment()
            } else if (oldItem is Group && newItem is Group) {
                typeContentTheSame = oldItem.numberOfChildEntries == newItem.numberOfChildEntries
            }
            return typeContentTheSame
                    && oldItem.nodeId == newItem.nodeId
                    && oldItem.type == newItem.type
                    && oldItem.title == newItem.title
                    && oldItem.icon == newItem.icon
                    && oldItem.isCurrentlyExpires == newItem.isCurrentlyExpires
        }

        override fun areItemsTheSame(item1: Node, item2: Node): Boolean {
            return item1 == item2
        }
    }

    fun contains(node: Node): Boolean {
        return mNodeSortedList.indexOf(node) != SortedList.INVALID_POSITION
    }

    /**
     * Add a node to the list
     * @param node Node to add
     */
    fun addNode(node: Node) {
        mNodeSortedList.add(node)
    }

    /**
     * Add nodes to the list
     * @param nodes Nodes to add
     */
    fun addNodes(nodes: List<Node>) {
        mNodeSortedList.addAll(nodes)
    }

    /**
     * Remove a node in the list
     * @param node Node to delete
     */
    fun removeNode(node: Node) {
        mNodeSortedList.remove(node)
    }

    /**
     * Remove nodes in the list
     * @param nodes Nodes to delete
     */
    fun removeNodes(nodes: List<Node>) {
        nodes.forEach { node ->
            mNodeSortedList.remove(node)
        }
    }

    /**
     * Remove a node at [position] in the list
     */
    fun removeNodeAt(position: Int) {
        mNodeSortedList.removeItemAt(position)
        // Refresh all the next items
        notifyItemRangeChanged(position, mNodeSortedList.size() - position)
    }

    /**
     * Remove nodes in the list by [positions]
     * Note : algorithm remove the higher position at each iteration
     */
    fun removeNodesAt(positions: IntArray) {
        val positionsSortDescending = positions.toMutableList()
        positionsSortDescending.sortDescending()
        positionsSortDescending.forEach {
            removeNodeAt(it)
        }
    }

    /**
     * Update a node in the list
     * @param oldNode Node before the update
     * @param newNode Node after the update
     */
    fun updateNode(oldNode: Node, newNode: Node) {
        mNodeSortedList.beginBatchedUpdates()
        mNodeSortedList.remove(oldNode)
        mNodeSortedList.add(newNode)
        mNodeSortedList.endBatchedUpdates()
    }

    /**
     * Update nodes in the list
     * @param oldNodes Nodes before the update
     * @param newNodes Node after the update
     */
    fun updateNodes(oldNodes: List<Node>, newNodes: List<Node>) {
        mNodeSortedList.beginBatchedUpdates()
        oldNodes.forEach { oldNode ->
            mNodeSortedList.remove(oldNode)
        }
        mNodeSortedList.addAll(newNodes)
        mNodeSortedList.endBatchedUpdates()
    }

    fun indexOf(node: Node): Int {
        return mNodeSortedList.indexOf(node)
    }

    fun notifyNodeChanged(node: Node) {
        notifyItemChanged(mNodeSortedList.indexOf(node))
    }

    fun setActionNodes(actionNodes: List<Node>) {
        this.mActionNodesList.apply {
            clear()
            addAll(actionNodes)
        }
    }

    fun unselectActionNodes() {
        mActionNodesList.forEach {
            notifyItemChanged(mNodeSortedList.indexOf(it))
        }
        this.mActionNodesList.apply {
            clear()
        }
    }

    /**
     * Notify a change sort of the list
     */
    fun notifyChangeSort(sortNodeEnum: SortNodeEnum,
                         sortNodeParameters: SortNodeEnum.SortNodeParameters) {
        this.mNodeComparator = sortNodeEnum.getNodeComparator(database, sortNodeParameters)
    }

    override fun getItemViewType(position: Int): Int {
        return mNodeSortedList.get(position).type.ordinal
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NodeViewHolder {
        val view: View = if (viewType == Type.GROUP.ordinal) {
            mInflater.inflate(R.layout.item_list_nodes_group, parent, false)
        } else {
            mInflater.inflate(R.layout.item_list_nodes_entry, parent, false)
        }
        val nodeViewHolder = NodeViewHolder(view)
        mInfoTextDefaultDimension = nodeViewHolder.text.textSize
        mSubtextDefaultDimension = nodeViewHolder.subText.textSize
        nodeViewHolder.numberChildren?.let {
            mNumberChildrenTextDefaultDimension = it.textSize
        }
        return nodeViewHolder
    }

    override fun onBindViewHolder(holder: NodeViewHolder, position: Int) {
        val subNode = mNodeSortedList.get(position)

        // Node selection
        holder.container.isSelected = mActionNodesList.contains(subNode)

        // Assign image
        val iconColor = if (holder.container.isSelected)
            mContentSelectionColor
        else when (subNode.type) {
            Type.GROUP -> mIconGroupColor
            Type.ENTRY -> mIconEntryColor
        }
        holder.imageIdentifier?.setColorFilter(iconColor)
        holder.icon.apply {
            database.iconDrawableFactory.assignDatabaseIcon(this, subNode.icon, iconColor)
            // Relative size of the icon
            layoutParams?.apply {
                height = (mIconDefaultDimension * mPrefSizeMultiplier).toInt()
                width = (mIconDefaultDimension * mPrefSizeMultiplier).toInt()
            }
        }

        // Assign text
        holder.text.apply {
            text = subNode.title
            setTextSize(mTextSizeUnit, mInfoTextDefaultDimension, mPrefSizeMultiplier)
            strikeOut(subNode.isCurrentlyExpires)
        }
        // Add subText with username
        holder.subText.apply {
            text = ""
            strikeOut(subNode.isCurrentlyExpires)
            visibility = View.GONE
        }
        // Add meta text to show UUID
        holder.meta.apply {
            text = subNode.nodeId.toString()
            visibility = if (mShowUUID) View.VISIBLE else View.GONE
        }

        // Specific elements for entry
        if (subNode.type == Type.ENTRY) {
            val entry = subNode as Entry
            database.startManageEntry(entry)

            holder.text.text = entry.getVisualTitle()
            holder.subText.apply {
                val username = entry.username
                if (mShowUserNames && username.isNotEmpty()) {
                    visibility = View.VISIBLE
                    text = username
                    setTextSize(mTextSizeUnit, mSubtextDefaultDimension, mPrefSizeMultiplier)
                }
            }

            val otpElement = entry.getOtpElement()
            holder.otpContainer?.removeCallbacks(holder.otpRunnable)
            if (otpElement != null
                && mShowOTP
                && otpElement.token.isNotEmpty()) {

                // Execute runnable to show progress
                holder.otpRunnable.action = {
                    populateOtpView(holder, otpElement)
                }
                if (otpElement.type == OtpType.TOTP) {
                    holder.otpRunnable.postDelayed()
                }
                populateOtpView(holder, otpElement)

                holder.otpContainer?.visibility = View.VISIBLE
            } else {
                holder.otpContainer?.visibility = View.GONE
            }
            holder.attachmentIcon?.visibility =
                    if (entry.containsAttachment()) View.VISIBLE else View.GONE

            database.stopManageEntry(entry)
        }

        // Add number of entries in groups
        if (subNode.type == Type.GROUP) {
            if (mShowNumberEntries) {
                holder.numberChildren?.apply {
                    text = (subNode as Group)
                            .numberOfChildEntries
                            .toString()
                    setTextSize(mTextSizeUnit, mNumberChildrenTextDefaultDimension, mPrefSizeMultiplier)
                    visibility = View.VISIBLE
                }
            } else {
                holder.numberChildren?.visibility = View.GONE
            }
        }

        // Assign click
        holder.container.setOnClickListener {
            mNodeClickCallback?.onNodeClick(database, subNode)
        }
        holder.container.setOnLongClickListener {
            mNodeClickCallback?.onNodeLongClick(database, subNode) ?: false
        }
    }

    private fun populateOtpView(holder: NodeViewHolder?, otpElement: OtpElement?) {
        when (otpElement?.type) {
            OtpType.HOTP -> {
                holder?.otpProgress?.apply {
                    max = 100
                    progress = 100
                }
            }
            OtpType.TOTP -> {
                holder?.otpProgress?.apply {
                    max = otpElement.period
                    progress = otpElement.secondsRemaining
                }
            }
        }
        holder?.otpToken?.text = otpElement?.token
        holder?.otpContainer?.setOnClickListener {
            otpElement?.token?.let { token ->
                Toast.makeText(
                        context,
                        context.getString(R.string.copy_field,
                                TemplateField.getLocalizedName(context, TemplateField.LABEL_TOKEN)),
                        Toast.LENGTH_LONG
                ).show()
                mClipboardHelper.copyToClipboard(token)
            }
        }
    }

    class OtpRunnable(val view: View?): Runnable {

        var action: (() -> Unit)? = null

        override fun run() {
            action?.invoke()
            postDelayed()
        }

        fun postDelayed() {
            view?.postDelayed(this, 1000)
        }
    }

    override fun getItemCount(): Int {
        return mNodeSortedList.size()
    }

    /**
     * Assign a listener when a node is clicked
     */
    fun setOnNodeClickListener(nodeClickCallback: NodeClickCallback?) {
        this.mNodeClickCallback = nodeClickCallback
    }

    /**
     * Callback listener to redefine to do an action when a node is click
     */
    interface NodeClickCallback {
        fun onNodeClick(database: Database, node: Node)
        fun onNodeLongClick(database: Database, node: Node): Boolean
    }

    class NodeViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var container: View = itemView.findViewById(R.id.node_container)
        var imageIdentifier: ImageView? = itemView.findViewById(R.id.node_image_identifier)
        var icon: ImageView = itemView.findViewById(R.id.node_icon)
        var text: TextView = itemView.findViewById(R.id.node_text)
        var subText: TextView = itemView.findViewById(R.id.node_subtext)
        var meta: TextView = itemView.findViewById(R.id.node_meta)
        var otpContainer: ViewGroup? = itemView.findViewById(R.id.node_otp_container)
        var otpProgress: ProgressBar? = itemView.findViewById(R.id.node_otp_progress)
        var otpToken: TextView? = itemView.findViewById(R.id.node_otp_token)
        var otpRunnable: OtpRunnable = OtpRunnable(otpContainer)
        var numberChildren: TextView? = itemView.findViewById(R.id.node_child_numbers)
        var attachmentIcon: ImageView? = itemView.findViewById(R.id.node_attachment_icon)
    }

    companion object {
        private val TAG = NodeAdapter::class.java.name
    }
}
